### PR TITLE
Update x509 tool to write all output to common BIO which is a file or stdout

### PR DESCRIPTION
**Pull request by andrewhop**
### Description of changes: 
Previously only some output followed the user's request using `-out` to direct the output to a file. This change updates everything to use a common `BIO` which is `stdout` by default or the user provided file.

### Testing:
Added new tests that test writing to a file directly with `-out` we have existing tests that cover writing to `stdout` (and then getting redirected to a file with `>`) works as expected.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: tool-openssl/x509.cc [hunk 1]
```diff
@@ -279,15 +272,8 @@ bool X509Tool(const args_list_t &args) {
...
-    if (!out_path.empty()) {
-      if (!WriteSignedCertificate(x509.get(), out_path)) {
-        return false;
-      }
-    }
-
-    if (!noout && !in_path.empty() && !checkend && parsed_args.count("-out")==0) {
-      bssl::UniquePtr<BIO> bio_out(BIO_new_fp(stdout, BIO_NOCLOSE));
-      if (!PEM_write_bio_X509(bio_out.get(), x509.get())) {
+    if (!noout && !checkend && !modulus) {
```
Comment (WillChilds-Klein) [inline tool-openssl/x509.cc:275]: we didn't check `modulus` previously here. why do it now?

Comment (andrewhop) [reply tool-openssl/x509.cc]: I was trying to fix a failing test, turns out OpenSSL 1.0.2 and 1.1.1 have wildly different behavior of what gets written to the -out parameter vs what is always written to stdout. I think testing against 1.0.2 has very limited benefit and is hurting our testing against newer versions.

**Review by WillChilds-Klein — DISMISSED** [review]

File: tool-openssl/x509.cc [hunk 2]
```diff
@@ -22,13 +22,8 @@ static const argument_t kArguments[] = {
...
-bool WriteSignedCertificate(X509 *x509, const std::string &out_path) {
-  ScopedFILE out_file(fopen(out_path.c_str(), "wb"));
-  if (!out_file) {
-    fprintf(stderr, "Error: unable to open output file '%s'\n", out_path.c_str());
-    return false;
-  }
-  if (!PEM_write_X509(out_file.get(), x509)) {
+static bool WriteSignedCertificate(X509 *x509, bssl::UniquePtr<BIO> &output_bio, const std::string &out_path) {
+  if (!PEM_write_bio_X509(output_bio.get(), x509)) {
     fprintf(stderr, "Error: error writing certificate to '%s'\n", out_path.c_str());
```
[the remark this answers is not in the document]
Comment (andrewhop) [reply tool-openssl/x509.cc:27]: Good catch, I didn't think about this for this change. Turns out it is the empty string which we pass in as the default value in `GetString(&out_path, "-out", "", parsed_args);` I'll update that so it makes a bit more sense.

**Review by samuel40791765 — APPROVED** [review]

File: tool-openssl/x509.cc [hunk 3]
```diff
@@ -22,13 +22,8 @@ static const argument_t kArguments[] = {
...
 };
 
-bool WriteSignedCertificate(X509 *x509, const std::string &out_path) {
-  ScopedFILE out_file(fopen(out_path.c_str(), "wb"));
-  if (!out_file) {
-    fprintf(stderr, "Error: unable to open output file '%s'\n", out_path.c_str());
-    return false;
-  }
-  if (!PEM_write_X509(out_file.get(), x509)) {
+static bool WriteSignedCertificate(X509 *x509, bssl::UniquePtr<BIO> &output_bio, const std::string &out_path) {
```
Comment (samuel40791765) [inline tool-openssl/x509.cc:25]: Can `output_bio` be `const`?

**Review by WillChilds-Klein — DISMISSED** [review]

**Review by samuel40791765 — APPROVED** [review]

**Review by WillChilds-Klein — APPROVED** [review]

**Pull request merged**